Changes 1559 to have a single shared pool for all transactions. (#2924)
ethereum/EIPs: EIPS/eip-1559.md

There is a base fee per gas in protocol, which can move up or down each block according to a formula which is a function of gas used in parent block and gas target (formerly known as gas limit) of parent block.
The algorithm results in the base fee per gas increasing when blocks are above the gas target, and decreasing when blocks are below the gas target.
The base fee per gas is burned.

> Transactions specify both the maximum network fee per gas (aka: base fee) they are willing to pay, as well as the maximum fee per gas they are willing to give to miners to incentivize them to include their transaction (aka: miner bribe).

Reviewer: `FEECAP` was previously the absolute maximum that the transaction sender is willing to pay. I think setting the overall maximum and the miner bribe maximum (as previously specified) makes more sense, otherwise the user has to add the two together to work out what the overall maximum fee is.

Author: There was no intention to change the behavior here (this particular change was meant to be non-normative).  Do you have a suggestion for better wording than what I have that would make that more clear (ideally without going *too* in depth)?

Reviewer: Might be easier to specify the other way around - "transactions specify the maximum fee per gas they are willing to give to miners to incentivize them to include their transaction (aka: miner bribe), they also specify a maximum overall fee per gas they are willing to pay, including both the miner bribe and the network fee (aka: base fee)"

Author: ```suggestion
Transactions specify the maximum fee per gas they are willing to give to miners to incentivize them to include their transaction (aka: miner bribe).
Transactions also specify the maximum fee per gas they are willing to pay total (aka: fee cap), which covers both the miner bribe and the block's network fee per gas (aka: base fee).
```
How does this sound @bluepintail

	code_hash: int = 0

INITIAL_FORK_BLOCK_NUMBER = 10 # TBD
BASE_FEE_MAX_CHANGE_DENOMINATOR = 8
ELASTICITY_MULTIPLIER = 2

class World(ABC):
	def validate_block(self, block: Block) -> None:
		parent_base_fee = self.parent(block).base_fee
		parent_gas_used = self.parent(block).gas_used
		parent_gas_target = self.parent(block).gas_target
		transactions = self.transactions(block)

		# check if the block used too much gas
		assert block.gas_used <= block.gas_target * ELASTICITY_MULTIPLIER, 'invalid block: too much gas used'

		# check if the block changed the gas target too much
		assert block.gas_target <= parent_gas_target + parent_gas_target // 1024, 'invalid block: gas target increased too much'
		assert block.gas_target >= parent_gas_target - parent_gas_target // 1024, 'invalid block: gas target decreased too much'

		# check if the base fee is correct
		if parent_gas_used == parent_gas_target:
			expected_base_fee = parent_base_fee
		elif parent_gas_used > parent_gas_target:
			gas_delta = parent_gas_used - parent_gas_target

> 			fee_delta = max(parent_base_fee * gas_delta // parent_gas_target // BASE_FEE_MAX_CHANGE_DENOMINATOR, 1)

Reviewer 2: max not needed here, it is fine to move by near 0 wei gas prices

			expected_base_fee = parent_base_fee + fee_delta
		else:
			gas_delta = parent_gas_target - parent_gas_used
			fee_delta = parent_base_fee * gas_delta // parent_gas_target // BASE_FEE_MAX_CHANGE_DENOMINATOR

> 			expected_base_fee = parent_base_fee - fee_delta

Reviewer 2: need to address cases when expected_base_fee ends up being < 0

Author: Discussed this in Discord, recording result here for thoroughness: In that particular conditional branch, there is no way for `gas_delta` to be negative, thus there is no way for `fee_delta` to be negative.  The same is true for the other branch.

	code_hash: int = 0

INITIAL_FORK_BLOCK_NUMBER = 10 # TBD
BASE_FEE_MAX_CHANGE_DENOMINATOR = 8

> ELLASTICITY_MULTIPLIER = 2

Reviewer 3: ```suggestion
ELASTICITY_MULTIPLIER = 2
```

class World(ABC):
	def validate_block(self, block: Block) -> None:
		parent_base_fee = self.parent(block).base_fee
		parent_gas_used = self.parent(block).gas_used
		parent_gas_target = self.parent(block).gas_target
		transactions = self.transactions(block)

		# check if the block used too much gas

> 		assert block.gas_used <= block.gas_target * ELLASTICITY_MULTIPLIER, 'invalid block: too much gas used'

Reviewer 3: ```suggestion
		assert block.gas_used <= block.gas_target * ELASTICITY_MULTIPLIER, 'invalid block: too much gas used'
```
